pkg/dataobj/internal/dataset/predicate_order.go

```diff
@@ -0,0 +1,214 @@
+package dataset
+
+import (
+	"sort"
+)
+
+// SelectivityScore represents how selective a predicate is expected to be.
+// Lower scores mean more selective (fewer matching rows).
+type SelectivityScore float64
+
+const (
+	noMatchSelectivity  SelectivityScore = 0.0 // No match
+	matchAllSelectivity SelectivityScore = 1.0 // Matches all rows
+)
+
+// OrderPredicates orders the predicates based on their selectivity and cost as a simple heuristic.
+// - Lower selectivity (more filtering) is better
+// - Lower cost of processing a row is better
+//
+// TODO: a permutation-based approach would be more accurate as it would account
+// for the cost of processing a predicate and how a predicate's selectivity affects
+// the row count for subsequent predicates reducing their processing cost.
+func OrderPredicates(predicates []Predicate) []Predicate {
+	if len(predicates) <= 1 {
+		return predicates
+	}
+
+	sort.Slice(predicates, func(i, j int) bool {
+		s1 := getPredicateSelectivity(predicates[i])
+		s2 := getPredicateSelectivity(predicates[j])
+
+		// for selectivity of 0, we can assume there is no cost to reading the column since the reader will prune the rows
+		// return early in such cases to avoid cost calculations
+		if s1 == 0 {
+			return true
+		} else if s2 == 0 {
+			return false
+		}
+
+		return float64(s1)*float64(getRowEvaluationCost(predicates[i])) < float64(s2)*float64(getRowEvaluationCost(predicates[j]))
+	})
+
+	return predicates
+}
+
+// getPredicateSelectivity returns a selectivity score representing the estimated percentage
+// of rows that will match (0.0 to 1.0). Lower scores mean more selective (fewer matching rows).
+func getPredicateSelectivity(p Predicate) SelectivityScore {
+	if p == nil {
+		return matchAllSelectivity
+	}
+
+	switch p := p.(type) {
+	case EqualPredicate:
+		var (
+			minValue, maxValue Value
+			info               = p.Column.ColumnInfo()
+		)
+
+		if e1, e2 := minValue.UnmarshalBinary(info.Statistics.MinValue), maxValue.UnmarshalBinary(info.Statistics.MaxValue); e1 == nil && e2 == nil {
+			if CompareValues(p.Value, minValue) < 0 || CompareValues(p.Value, maxValue) > 0 {
+				// no rows will match
+				return noMatchSelectivity
+			}
+		}
+
+		// For equality, estimate rows per unique value
+		// selectivity = rows per unique value / total rows
+		matchingRows := float64(info.ValuesCount) / float64(info.Statistics.CardinalityCount)
```

Comment:
I think cardinality count can be 0 / nil for some columns which would end up panicing

Reply:
handled in https://github.com/grafana/loki/pull/17376/commits/b4997d8236b72329b8014c957ac56ffddbd7f36c